File: metrics.py
import numpy as np
import pandas as pd

class PerformanceMetrics:
    '''Utility class for evaluating risk-adjusted performance metrics.'''
    @staticmethod
    def calculate_sharpe_ratio(portfolio_values, risk_free_rate=0.0, periods_per_year=252):
        '''Calculates annualised Sharpe Ratio based on daily portfolio returns.'''
        equity_series = pd.Series(portfolio_values)
        daily_returns = equity_series.pct_change().dropna() # Vectorised operation applied to whole data list
        if daily_returns.std() == 0:
            return 0.0
        
        # Excess returns over risk-free rate per period
        mean_daily_return = daily_returns.mean() - (risk_free_rate / periods_per_year)
        std_daily_return = daily_returns.std()

        # Annualise the daily ratio by multiplying by sqrt(252 trading days)
        sharpe_ratio = (mean_daily_return / std_daily_return) * np.sqrt(periods_per_year)
        return sharpe_ratio
    
    @staticmethod
    def calculate_max_drawdown(portfolio_values):
        '''Calculates Max Drawdown (MDD) percentage from largest peak-to-trough drop in value of investment portfolio.'''
        equity_series = pd.Series(portfolio_values)
        running_max = equity_series.cummax() # Rolling peak across timeline, another vectorised operation
        drawdowns = (equity_series - running_max) / running_max # Percentage drawdown from rolling peak
        max_drawdown = drawdowns.min() * 100
        return max_drawdown

File: Metrics.cpp
#include "Metrics.h"
#include <algorithm>
#include <cmath> // Standard maths operations
#include <numeric> // For std::accumulate - fast array algorithms

using namespace std;

double PerformanceMetrics::calculateSharpeRatio(const vector<double>& portfolio_values,
                                                double risk_free_rate,
                                                int periods_per_year) {
    if (portfolio_values.size() < 2) return 0.0; // Need at least 2 daily equity values to calculate percentage return

    // Calculate daily percentage returns
    vector<double> daily_returns;
    daily_returns.reserve(portfolio_values.size() - 1); // Pre-allocates space in memory!
    for (size_t i = 1; i < portfolio_values.size(); ++i) {
        double ret = (portfolio_values[i] - portfolio_values[i - 1]) / portfolio_values[i - 1]; // Standard percentage change
        daily_returns.push_back(ret);
    }

    // Calculate mean daily return
    double sum = accumulate(daily_returns.begin(), daily_returns.end(), 0.0); // Iterates from start to end, summing every number, starting from 0.0
    double mean_return = sum / daily_returns.size();

    // Calculate daily return standard deviation - we want to know how much the returns deviate from the mean return, which is a measure of risk.
    double variance_sum = 0.0;
    for (double r : daily_returns) {
        variance_sum += (r - mean_return) * (r - mean_return); // Sum of squared deviations from the mean
    }

    // For sample standard deviation, divide by (n-1)
    double std_dev = sqrt(variance_sum / (daily_returns.size() - 1));
    if (std_dev == 0.0) return 0.0; // Avoid division by zero

    // Annualise Sharpe Ratio - (Excess Return / Standard Deviation) * sqrt(periods_per_year)
    double excess_return = risk_free_rate / periods_per_year;
    double sharpe_ratio = (mean_return - excess_return) / std_dev * sqrt(periods_per_year);

    return sharpe_ratio;
                                                }

double PerformanceMetrics::calculateMaxDrawdown(const vector<double>& portfolio_values) {
    if (portfolio_values.empty()) return 0.0; // Guard clause for empty portfolio values
    double max_drawdown = 0.0;
    double running_peak = portfolio_values[0]; // Initialise running peak to Day 1 value
    
    // Track running peak and maximum negative deviation
    for (double current_value : portfolio_values) {
        if (current_value > running_peak) {
            running_peak = current_value;
        }
        double drawdown = (current_value - running_peak) / running_peak; // Calculate drawdown as a percentage
        if (drawdown < max_drawdown) {
            max_drawdown = drawdown; // Update max drawdown if current drawdown is worse
        }
    }
    return max_drawdown * 100.0;  
}